File: Data_Structure/Union_Find.cpp
struct UnionFind{
	std::vector<int> data;
	
	UnionFind(int n){
		data.assign(n, -1);
	}
	
	bool unite(int x, int y){
		x = find(x); y = find(y);
		if(x == y) return false;
		if(data[x] > data[y]) std::swap(x, y);
		data[x] += data[y];
		data[y] = x;
		return true;
	}
	
	bool same(int x, int y){
		return find(x) == find(y);
	}
	
	int find(int k){
		if(data[k] < 0) return k;
		return data[k] = find(data[k]);
	}
	
	int size(int k){
		return -data[find(k)];
	}
};

File: Rust/Algorithm/UnionFind.rs
struct UnionFind {
	data: Vec<i32>,
}

impl UnionFind {
	fn new(n: usize) -> UnionFind {
		UnionFind {
			data: vec![-1; n],
		}
	}
	
	fn unite(&mut self, x: usize, y: usize) -> bool {
		let mut x = self.find(x);
		let mut y = self.find(y);
		if x == y { return false; }
		if self.data[x] > self.data[y] { std::mem::swap(&mut x, &mut y); }
		self.data[x] += self.data[y];
		self.data[y] = x as i32;
		true
	}
	
	fn find(&mut self, k: usize) -> usize {
		if self.data[k] < 0 {
			k
		}else{
			let l = self.data[k];
			self.data[k] = self.find(l as usize) as i32;
			self.data[k] as usize
		}
	}
	
	fn same(&mut self, x: usize, y: usize) -> bool {
		self.find(x) == self.find(y)
	}
	
	fn size(&mut self, k: usize) -> usize {
		let k = self.find(k);
		-self.data[k] as usize
	}
}
